Simplified top level of SMARTS

RutvikGupta opened the pull request.

RutvikGupta commented:
1. I decided not to move ./cli, ./envision and ./zoo to ./smarts since they vary differently and it makes sense to keep them separate.

2. I read some articles and checked out popular python repositories and they all recommended to keep requirements.txt and setup.py at the top level as a good repo structure practice. 

3. I moved test_pip_packages.sh to etc/setup folder which resulted in CI Format & Dependencies / test-requirements failing even though I corrected the path at .yml file. Its not able to access the file and gives a permission denied error. I am not sure I can fix this but I think the remote on which these CI tests are running does not have the permission the access the folder I created. 

4. The docker file path change also involved changes to README.md, so I recommend to check that out as well.

5.  I renamed benchmark to marl_benchmark as it is the Multiagent-benchmark cases. I put this into a cases folder as it is a use case like Ultra. Should I move ultra to this folder as well. ?

Adaickalavan commented on README.md:
````diff
@@ -238,11 +238,12 @@ python examples/run_smarts.py --algo SAC --scenario ./scenarios/loop --n_agents
 
 ```bash
-$ cd /path/to/SMARTS
+$ cd /path/to/SMARTS/etc/docker
````
```suggestion
$ cd /path/to/SMARTS
```
Change in directory is not needed here.

Adaickalavan commented on README.md:
````diff
@@ -263,6 +264,7 @@ $ python examples/single_agent.py scenarios/loop
 
 ```bash
 # For this to work, your account needs to be added to the huaweinoah org
+$ cd /path/to/SMARTS/etc/docker
````
```suggestion
```

Change in directory is not needed here.

RutvikGupta replied:
Thanks for that. I changed it back.

Adaickalavan commented on etc/setup/README.pypi.md:
```diff
@@ -0,0 +1,227 @@
+# SMARTS
```
Why is `README.pypi.md` file present in two locations, namely, `SMARTS/etc/setup/` and `SMARTS/` ?

Adaickalavan commented on README.md:
```diff
@@ -243,6 +243,7 @@ $ docker run --rm -it -v $PWD:/src -p 8081:8081 huaweinoah/smarts:<version>
 # E.g. docker run --rm -it -v $PWD:/src -p 8081:8081 huaweinoah/smarts:v0.4.12
 # <press enter>
 
+$ cd ../..
```
Change in directory not needed here.

Adaickalavan commented on README.md:
````diff
@@ -263,6 +264,7 @@ $ python examples/single_agent.py scenarios/loop
 
 ```bash
 # For this to work, your account needs to be added to the huaweinoah org
+$ cd /path/to/SMARTS
````
```suggestion
$ cd /path/to/SMARTS
export VERSION=v0.4.16
docker build --no-cache -f ./etc/docker/Dockerfile -t smarts:$VERSION .
docker tag smarts:$VERSION huaweinoah/smarts:$VERSION
docker login
docker push huaweinoah/smarts:$VERSION
```

Adaickalavan commented on CHANGELOG.md:
```diff
@@ -22,9 +22,10 @@ process
 - Added `--allow-offset-map` option for `scl scenario build` to prevent auto-shifting of Sumo road networks
 - Added options in `DoneCriteria` to trigger ego agent to be done based on other agent's done situation
-- Added README_pypi.md to update to the general user installation PyPI instructions. See Issue #828 
+- Added `README_pypi.md` to update to the general user installation PyPI instructions. See Issue #828 
 ### Changed
 - Refactored SMARTS class to not inherit from Panda3D's ShowBase; it's aggregated instead. See issue #597.
+- Refactored the top level of the SMARTS module to make it easier to navigate the project and understand its structure. See issue #776.
```
These additions and changes should be moved to the appropriate heading in the `Unreleased` section at the top. Currently, they are listed under release 0.4.16 section.

Adaickalavan replied:
The changes should be moved to the `Changed` heading in the `Unreleased` section at the top. Currently, it is listed under release `0.4.16` section.

Adaickalavan commented on etc/setup/requirements.txt:
```diff
@@ -0,0 +1,159 @@
+absl-py==0.10.0
```
Why is the `requirements.txt` file present in two locations, namely, `SMARTS/etc/setup/` and `SMARTS/` ?

RutvikGupta replied:
I was checking whether moving the requirements file would fix the CI test failure. I removed the requirements.txt from etc/setup.

Adaickalavan commented on etc/docker/Dockerfile:
```diff
@@ -52,12 +52,12 @@ RUN wget -O /etc/X11/xorg.conf http://xpra.org/xorg.conf && \
 
 # Install requirements.txt
-COPY requirements.txt /tmp/requirements.txt
+COPY ../../requirements.txt /tmp/requirements.txt
```
```suggestion
COPY ./requirements.txt /tmp/requirements.txt
```

Change the exact location depending on whether `SMARTS/requirements.txt` or `SMARTS/etc/setup/requirements.txt` is used.

Adaickalavan commented on etc/docker/Dockerfile:
```diff
@@ -52,12 +52,12 @@ RUN wget -O /etc/X11/xorg.conf http://xpra.org/xorg.conf && \
 # Copy source files and install SMARTS
 ENV PYTHONPATH=/src
-COPY . /src
+COPY ../../. /src
```
```suggestion
COPY . /src
```

Adaickalavan commented:
https://github.com/huawei-noah/SMARTS/blob/bb2bf9cd143629ca89c4e03979c8c8ce1d70377a/etc/docker/Dockerfile.minimal#L5

```suggestion
# $ docker build -f ./etc/docker/Dockerfile.minimal -t huaweinoah/smarts:$VERSION-minimal .
```

RutvikGupta commented:
Thanks @Adaickalavan  for your review. 
About the formatting and testing requirement CI test that I am failing, I think its because of change in the filepath to dockerfile. Can u please help me with that. Its about this one,
3. I moved test_pip_packages.sh to etc/setup folder which resulted in CI Format & Dependencies / test-requirements failing even though I corrected the path at .yml file. Its not able to access the file and gives a permission denied error. I am not sure I can fix this but I think the remote on which these CI tests are running does not have the permission the access the folder I created.

Gamenot commented on .github/workflows/ci-formatting-and-dependencies.yml:
```diff
@@ -34,4 +34,4 @@ jobs:
           cd $GITHUB_WORKSPACE
-          ./test_pip_packages.sh
\ No newline at end of file
+          . ./test_pip_packages.sh
```
Yeah, I just got this at about the same time. The file is missing executable flags.

sah-huawei reviewed (commented):
>     1. I decided not to move ./cli, ./envision and ./zoo to ./smarts since they vary differently and it makes sense to keep them separate.

One option we might consider is adding a `./src/` folder at the top level and them moving these, as well as `smarts` (and possibly `third_party`?) into there.   Not saying it's better... just putting out options in the spirit of brainstorming.

>     5. I renamed benchmark to marl_benchmark as it is the Multiagent-benchmark cases. I put this into a cases folder as it is a use case like Ultra. Should I move ultra to this folder as well. ?

I don't know about Ultra (I think there's some historical reasons for not messing with it?) but if we did add the `src` folder, I would be tempted to move it there too. 

 However, for the benchmark, I'm curious about how `cases` is different from `examples`.

RutvikGupta commented on cases/marl_benchmark/agents/networked_pg/tf_policy.py:
```diff
@@ -22,7 +22,7 @@
 from ray.rllib.policy.tf_policy_template import build_tf_policy
 
-from benchmark.networks.communicate import NetworkedMixin, postprocess_trajectory
+from etc.benchmark.networks.communicate import NetworkedMixin, postprocess_trajectory
```
yes you are right. I will change those.

RutvikGupta replied:
> One option we might consider is adding a `./src/` folder at the top level and them moving these, as well as `smarts` (and possibly `third_party`?) into there. Not saying it's better... just putting out options in the spirit of brainstorming.

I think, this is a good option but involves a major change to the structure and will break a bunch of things like packaging and editable (`pip -e`) use. @Gamenot suggests this should be deferred to the next major version. So i will open another issue which will address this change for now.

RutvikGupta commented:
> > ```
> > 1. I decided not to move ./cli, ./envision and ./zoo to ./smarts since they vary differently and it makes sense to keep them separate.
> > ```
> 
> One option we might consider is adding a `./src/` folder at the top level and them moving these, as well as `smarts` (and possibly `third_party`?) into there. Not saying it's better... just putting out options in the spirit of brainstorming.
> 
> > ```
> > 5. I renamed benchmark to marl_benchmark as it is the Multiagent-benchmark cases. I put this into a cases folder as it is a use case like Ultra. Should I move ultra to this folder as well. ?
> > ```
> 
> I don't know about Ultra (I think there's some historical reasons for not messing with it?) but if we did add the `src` folder, I would be tempted to move it there too.
> 
> However, for the benchmark, I'm curious about how `cases` is different from `examples`.

I moved becnhmark because it is a completely different case of SMARTS like ultra is. Examples are more general cases of SMARTS. I can move the examples to cases as well if you want.

sah-huawei commented on CONTRIBUTING.md:
```diff
@@ -32,9 +56,54 @@ Please take a look at the current open issues to see if any of them interest you
 5. Describe the problem and add references to the related issues that the request solves.
 6. Request review of your code by other contributors. Try to improve your code as much as possible to lessen the burden on others.
```
Should we also mention our convention about adding entries to `CHANGELOG.md` within pull requests?

RutvikGupta replied:
Yeah, that's a good point. I will add that.

sah-huawei commented on CONTRIBUTING.md:
```diff
@@ -32,9 +56,54 @@ Please take a look at the current open issues to see if any of them interest you
 5. Describe the problem and add references to the related issues that the request solves.
 6. Request review of your code by other contributors. Try to improve your code as much as possible to lessen the burden on others.
 
+## Merge Requests
```
We have a "Pull Requests" subsection just above this.  I personally don't distinguish between "Pull" and "Merge" requests.  Is there a difference?  If not, we should combine these two subsections.  If so, maybe we need a 1-liner to explain the distinction?  (Note that below, in the paragraph after the list, the term "pull request" is used twice.  That leads me to believe this subsection should be combined with the previous one.)

RutvikGupta replied:
I realized now that pull requests were called merge requests before. I will merge the two.

sah-huawei commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,15 @@
+## Code of Conduct
+- Be respectful. 
+- Keep criticism strictly to code.
+- Ask a project lead before revealing information about the project to outside entities. 
```
For open-source contributors, it's not obvious who the "project leads" are, nor what would constitute an "outside entity".

sah-huawei commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,15 @@
+- Be respectful. 
+- Keep criticism strictly to code.
+- Ask a project lead before revealing information about the project to outside entities. 
+- If in doubt about conduct ask.
```
whom?

RutvikGupta replied:
Other contributors. I will add that.

sah-huawei commented on CONTRIBUTING.md:
```diff
@@ -23,6 +31,22 @@ Once done, you're all set to make your first contribution!
 
+## Committing
+
+Please take care in using good commit messages as they're useful for debugging, reviewing code, and generally just shows care and quality. [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/) provides a good guideline. At a minimum,
```
While this is true in general, it's less important when grouping all commits into PRs that will be squashed when merged and the branches subsequently deleted.  But I guess we should still keep this section, if only for the `[ci skip]` instruction, which is useful.  (I always forget about that!)

sah-huawei commented on CONTRIBUTING.md:
```diff
@@ -32,9 +56,54 @@ Please take a look at the current open issues to see if any of them interest you
 5. Describe the problem and add references to the related issues that the request solves.
```
"addresses" instead of "solves".  (not all issues are bugs, nor do all PRs completely resolve an issue.)
